IDA* algorithm implementation

aandre-costaa opened the pull request:
IDA* algorithm implementation for networkx. It works with directed and undirected graphs and heuristics in a similar way to the A* algorithm. 

This algorithm works by creating a threshold for each iteration and analysing the nodes that have the minimum heuristic cost below the threshold value (if there's none below the threshold, it expands the less expensive one above that value). 

The algorithm works with a recursive search function which proved to be faster rather then an infinite cycle function. The search function analyses the node against the threshold and later it's neighbours to expand the most promising one. Also keeps track of the optimal path explored nodes only releasing the ones it finds not fit, this makes it lighter on memory in comparison to A*.

Recursive function:
![image_2025-04-09_150613818](https://github.com/user-attachments/assets/bc9021e8-103e-4e86-aeb8-bd938be1149c)

Non-recursive function:
![image_2025-04-09_150639619](https://github.com/user-attachments/assets/a5c3216e-aa9c-47a7-868e-fc90a7466c28)

rossbar commented:
See also #6896 , which proposed to add another extension of A*. Given the interest in this particular sub-domain, it'd be great if we could develop a more comprehensive understanding of the lay-of-the-land and figure out if/how/whether such extensions can be offered in a coherent way!

aandre-costaa commented:
Hi, thanks for reviewing. 
While in the development of the two functions tested the non recursive one was developed using a stack to store and manage neighbours to later analyse what node was the most promising one to expand, i was honestly unaware of how big python's recursive limitations were and so thought it was ok to develop it that way. 
I did think about the recursive being worse than just a cycle with stack and that's why i tested both, however after seeing that both used the same peak memory but the recursive one took half the time i assumed that would be the better approach (even while managing the stack to just contain the next set of nodes and poping it afterwards). What i can do is test them again with an even larger graph and see if the difference in performance of both starts changing.

Also, will try to fit my algorithm in the astar.py module and add the doc_strings as soon as i can.

When it comes @rossbar 's suggestion, since these algorithms are only evolutions of the A* algorithm with some slight changes on the way ir handles nodes, could it be possible to deliver these algorithms as sub-modules to the A*? Something like having the A* module/algorithm and then sub-modules with it's evolutions?

aandre-costaa commented:
Did some testing on the code and yes, you were correct, when in a graph with more than 1000 nodes the recursion limit is hit. 

Went back to the code of the non recursive function and improved it, it is a stack based function and i improved the way the stack is managed to only contain the necessary nodes on each iteration. Also tested another non recursive function adapted to use a list instead of the stack but the stack proved to be better in consistency and memory.

The algorithm is now non recursive and uses the same memory as it did with recursion while keeping pretty much the same time to find the optimal path.

Added the doc_strings as asked and moved the code to the ```astar.py``` module

List based function
![image](https://github.com/user-attachments/assets/5054652a-f84f-4dd8-9a9b-cdbcdd1823f7)


Stack improved function
![image](https://github.com/user-attachments/assets/0739aeb4-423a-4199-bba7-48d96b370883)

aandre-costaa commented:
Hello again, on another important note i must inform that I am apart of a student group project that in cooperation with our course coordinator is taking the task of investigating path-finding algorithms, specifically the evolutions of the A* algorithm and contributing with the algorithms developed to the Networkx library to try and expand the algorithms offered by it and contribute to the community. 
This means that the current algorithms that are being implemented in the past couple of days are being developed by this student group, myself, included as part of the project, and in the next couple of days it is predicted to have one or two more algorithms added to the pull requests.

dschult commented:
Thank you for reporting about the class project. I figured that must be what is going on. 

Could you ask your professor to consider combining these approaches into a unified approach to astar variants? Or maybe have them open an Issue to coordinate the discussion of astar algorithms in NetworkX. Thanks to the professor and to you!

rossbar commented:
Just wanted to ping on the A* projects - are folks still working on them or is the semester over?

My personal opinion is that absent a strong, dedicated "A* champion" (or champions) who is/are willing to review these and engage with the NetworkX community, these are unlikely to move forward. Bear in mind that any algorithms that are added to the library become the responsibility of the NetworkX community to maintain in the future. IMO the biggest blocker here is a lack of expertise in the existing pool of active reviewers/community members to feel comfortable making this promise.

Absent this level of engagement, I am inclined to close them and will label them as such until others have a chance to weigh in. Note that my doing so is in no way an indictment of code quality or a judgement that these are not valuable algorithms, but simply my personal judgement that at present time such additions would be unmaintainable. It would be wonderful if that were to change! If there's any interest at all in getting more involved in NetworkX, we would strongly encourage it! Looking at existing functionality and opening issues/PRs is a great way to familiarize oneself with the codebase and build ties with the community for future proposals. Checking out the [community meetings (scroll to the bottom for rendered calendar)](https://scientific-python.org/calendars/) is another great way to get involved if you can make it!

MridulS commented:
I'll go ahead and close these PRs, and just to reiterate @rossbar [comment](https://github.com/networkx/networkx/pull/7965#issuecomment-3053379335) that this is *not* about the code in the PRs or any value judgements about the algorithms. It's just a lot of new algorithms without any dedicated champions on the networkx community/maintainer side 😅

I would also suggest to package all the variants up in a new package that depends on networkx and that can be distributed via github/pypi. The scientific python cookie project should take care of all the templating https://github.com/scientific-python/cookie for a new python library! Feel free to ping/tag us if you need any help with this :)
